src/database/mod.rs

```diff
@@ -61,6 +129,34 @@ pub enum WorkflowType {
     External,
 }
 
+impl sqlx::Type<Postgres> for WorkflowType {
+    fn type_info() -> PgTypeInfo {
+        <String as sqlx::Type<Postgres>>::type_info()
+    }
+}
+
+impl sqlx::Decode<'_, Postgres> for WorkflowType {
```

Comment:
Maybe this could be done even easier with the `sqlx` proc macro https://docs.rs/sqlx/latest/sqlx/trait.Type.html#enumeration?

Reply:
I wish it could, but sqlx's support for postgres enum is quite limited. I had quite a difficult time attempting to query the BuildStatus enum within the BuildModel inside Workflow, only to be greeted by this issue: https://github.com/launchbadge/sqlx/issues/1920
> error occurred while decoding column variables: custom types in records are not fully supported yet: failed to retrieve type info for field 4 with type oid 25101253

So I just manually implemented `Decode` trait for each of those enums.

Reply:
I don't really want to use Postgres enums, just map Rust enums to strings in the DB. But I suppose that's not easily possible with sqlx (?).

Reply:
Yeah, I've just checked sqlx source code to be certain, and it appears under the hood `sqlx` use the database's native enum type to represent Rust enum.

Reply:
I checked the [docs](https://docs.rs/sqlx/latest/sqlx/trait.Type.html#enumeration) again and they claim that without `#[repr(i32)]` or `#[sqlx(type_name)]`, it should just be using text. It's not obvious from the documentation, but this works fine:
```rust
#[derive(sqlx::Type, Debug)]
#[sqlx(type_name = "TEXT")] // <- this is the important bit
#[sqlx(rename_all = "lowercase")]
enum Color {
    Red,
    Green,
    Blue
}

#[derive(sqlx::FromRow, Debug)]
struct Row {
    color: Color
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let pool = PgPool::connect("postgres://postgres:postgres@localhost/postgres").await?;
    sqlx::query("CREATE TABLE IF NOT EXISTS test(COLOR TEXT);").execute(&pool).await?;

    // DB contains { color: "red" }
    // sqlx::query("INSERT INTO test(COLOR) VALUES($1)").bind(Color::Red).execute(&pool).await?;
    let row: Row = sqlx::query_as("SELECT * FROM test").fetch_one(&pool).await?;
    println!("{row:?}");

    Ok(())
}
```
This would be ideal, however I haven't found a way how to get this working with `query!`, i.e. compile-time checked queries.

Reply:
Sorry, I wasn't clear earlier. What I meant was there was no way to use the enum within the query macro. I did not try the function version though

Reply:
It does work for inserts:
```rust
sqlx::query!("INSERT INTO test(color) VALUES($1)", Color::Red as Color).execute(&pool).await?;
```
Not sure how to query it back from the DB yet, though.

Reply:
Ok, this does the trick:
```rust
let row: Row = sqlx::query_as!(Row, r#"SELECT color AS "color!: Color" FROM test"#).fetch_one(&pool).await?;
```
it's a bit ugly, but at the same time it saves us from duplicating the encode/decode impls and the string constants within them.

Reply:
Wow, `type_name=TEXT` really cooked. Not once had I thought that I could put TEXT as type name, I kept trying to create a new Postgres enum instead 🤣 Thanks Jakub, now I know what to do the next time I encounter this issue.